minica/minica_test.go

```diff
@@ -0,0 +1,541 @@
+			CertType:        ssh.UserCert,
+			KeyId:           "jane@test.com",
+			ValidPrincipals: []string{"jane"},
+			ValidAfter:      uint64(time.Now().Unix()),
+			ValidBefore:     uint64(time.Now().Add(time.Hour).Unix()),
+		}}, ssh.UserCert, "jane", false},
+		{"ok infinity", mustCA(t), args{&ssh.Certificate{
+			Key:             publicKey,
+			Serial:          1234,
+			CertType:        ssh.UserCert,
+			KeyId:           "jane@test.com",
+			ValidPrincipals: []string{"jane"},
+			ValidBefore:     ssh.CertTimeInfinity,
+		}}, ssh.UserCert, "jane", false},
+		{"fail type", mustCA(t), args{&ssh.Certificate{
+			Key:             publicKey,
+			Serial:          1234,
+			CertType:        100,
+			KeyId:           "jane@test.com",
+			ValidPrincipals: []string{"jane"},
+		}}, 0, "", true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := tt.ca.SignSSH(tt.args.cert)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("CA.SignSSH() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+
+			if tt.wantErr {
+				if got != nil {
+					t.Errorf("CA.SignSSH() = %v, want nil", got)
+				}
+			} else {
+				checker := ssh.CertChecker{
+					IsUserAuthority: func(auth ssh.PublicKey) bool {
+						return reflect.DeepEqual(auth, tt.ca.SSHUserSigner.PublicKey())
+					},
+					IsHostAuthority: func(auth ssh.PublicKey, address string) bool {
+						return reflect.DeepEqual(auth, tt.ca.SSHHostSigner.PublicKey())
+					},
+				}
+				switch tt.wantCertType {
+				case ssh.HostCert:
+					if err := checker.CheckHostKey(tt.wantPrincipal+":22", &net.IPAddr{IP: net.IP{1, 2, 3, 4}}, got); err != nil {
+						t.Errorf("CertChecker.CheckHostKey() error = %v", err)
+					}
+				case ssh.UserCert:
+					if _, err := checker.Authenticate(mockConnMetadata(tt.wantPrincipal), got); err != nil {
+						t.Errorf("CertChecker.Authenticate() error = %v", err)
+					}
+				default:
+					t.Fatalf("unknown cert type %v", tt.wantCertType)
+				}
+			}
+		})
+	}
+}
+
+func TestCA_SignSSH_mutation(t *testing.T) {
+	signer, err := keyutil.GenerateDefaultSigner()
+	if err != nil {
+		t.Fatal(err)
+	}
+	publicKey, err := ssh.NewPublicKey(signer.Public())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	template := &ssh.Certificate{
+		Key:             publicKey,
+		CertType:        ssh.HostCert,
+		KeyId:           "ssh.test.com",
+		ValidPrincipals: []string{"ssh.test.com"},
+	}
+
+	ca := mustCA(t)
+	got, err := ca.SignSSH(template)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Validate certificate
+	checker := ssh.CertChecker{
+		IsHostAuthority: func(auth ssh.PublicKey, address string) bool {
+			return reflect.DeepEqual(auth, ca.SSHHostSigner.PublicKey())
+		},
+	}
+	if err := checker.CheckHostKey("ssh.test.com:22", &net.IPAddr{IP: net.IP{1, 2, 3, 4}}, got); err != nil {
+		t.Errorf("CertChecker.CheckHostKey() error = %v", err)
+	}
+
+	// Validate mutation
+	if template.ValidAfter != 0 {
+		t.Errorf("CA.SignSSH() mutated template.ValidAfter")
```

Comment:
there's a lot of `Errorf` calls in which you don't interpolate anything, I feel as if `t.Error` would be more appropriate in these cases.

Reply:
Fixed with 98fa65a